endaq/ide/nmea.py

```diff
@@ -0,0 +1,173 @@
+
+
+def get_nmea_sentences(dataset: [Dataset], channel=None, raw=False):
+    """ Processes an IDE file to return the parsed data.
+        Raise error if there is no raw NMEA channel.
+
+        :param dataset: a dataset, as from endaq.ide.files.get_doc()
+        :param channel: allows user to specify dataset channel
+        :param raw: if true, dump raw nmea sentences as opposed to parsed objects. Defaults to false
+        :return: NMEA sentence as NMEAMessage objects or bytes
+        :raise ValueError: if dataset is not a Dataset object
+    """
+    if not isinstance(dataset, Dataset):
+        raise ValueError("Requires Dataset object as input")
+    # fill element list with the needed data (given by david)
+    nmeaSentences = []
+    rawBuffer = bytearray()
+    nmea_channel_id = None
+    if channel is None:  # find NMEA channel
+        for chid, ch, in dataset.channels.items():
+            if "NMEA" in ch.name:
+                nmea_channel_id = chid
+                break
+    else:  # given NMEA channel
+        nmea_channel_id = channel
+    if nmea_channel_id not in dataset.channels:  # check we found a valid channel ID
+        raise ValueError("No raw NMEA channel specified or found")
+
+    for el in dataset.ebmldoc:
+        if el.name == "ChannelDataBlock":
+            if el[0].value == nmea_channel_id:
+                rawBuffer.extend(el.dump()["ChannelDataPayload"])
+                extractData, rawBuffer = get_nmea(rawBuffer, raw)
+                nmeaSentences.extend(extractData)
+    return nmeaSentences
+
+
+def get_nmea_measurement(data: [[NMEAMessage], Dataset], measure_type, filter_level: int = 0, timestamp="GPS") -> pd.DataFrame:
+    """ Returned dataframe format should appear like dataframes for other channels run through the
+        endaq.ide.to_pandas function.
+
+        :param data: A dataset or list of NMEAMessage Objects. The data to pull measurements from.
+        :param measure_type: Determines the kind of measurements that will be returned in the dataframe. Can accept
+            ANY (returns all following measurements), DIRECTION (degrees, True North), LOCATION (lat/long),
+            and SPEED (km/h)
+        :param filter_level: reliability filter level - Will only process data when connected to this number of
+            satellites. Defaults to 0, maximum of 12.
+        :param timestamp: Default is to provide timestamp from the GPS message. Selecting between GPS time and device
+            time will be the last piece of the job.
+        :raise ValueError: if data is not a list or does not contain NMEAMessage objects
+        :see: NMEA reference <https://www.sparkfun.com/datasheets/GPS/NMEA%20Reference%20Manual-Rev2.1-Dec07.pdf>`
+        :see: NMEAMessage Docs <https://github.com/semuconsulting/pynmeagps/blob/master/pynmeagps/nmeamessage.py>
+    """
+    if (not isinstance(data, list) or not isinstance(data[0], NMEAMessage)) and not isinstance(data, Dataset):
```

Comment:
This isn't going to work; if `data` is not a list, then `data[0]` is likely to raise a `TypeError`. Should probably be `if (not data or not (isinstance(data, list) and not isinstance(data[0], MNEAMessage)) and not isinstance(data, Dataset):`

Reply:
I think my example might be slightly wrong, but you get the idea. The initial `not data` will prevent a problem if `data` is `[]`